### PR TITLE
Use export map to restrict visible lib symbols under MinGW.

### PR DESCRIPTION
**Pull request by logzero**
This deals with the duplicate symbols issue.
https://github.com/DiligentGraphics/DiligentEngine/issues/31

### CODE REVIEW

File: Graphics/GraphicsEngineOpenGL/CMakeLists.txt [hunk 1]
```diff
@@ -213,10 +213,20 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
...
-        LINK_FLAGS "-Wl,--no-undefined -Wl,--no-allow-shlib-undefined"
-    )
+    if(PLATFORM_WIN32)
+        file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/export.map
+            "{ global: *GetEngineFactoryOpenGL*; local: *; };"
+        )
+        set_target_properties(Diligent-GraphicsEngineOpenGL-shared PROPERTIES
+            # Disallow missing direct and indirect dependencies to enssure that .so is self-contained
+            LINK_FLAGS "-Wl,--no-undefined -Wl,--no-allow-shlib-undefined -Wl,--version-script=export.map"
+        )
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineOpenGL/CMakeLists.txt:223]: Will this work if you just have this win32-specific branch to eliminate duplication:
```
if(PLATFORM_WIN32)
    # Please write a comment about what this does
    file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/export.map
          "{ global: *GetEngineFactoryOpenGL*; local: *; };"
    )
    set_target_properties(Diligent-GraphicsEngineOpenGL-shared PROPERTIES
           LINK_FLAGS "-Wl,--version-script=export.map"
    )
endif()
```

File: Graphics/GraphicsEngineVulkan/CMakeLists.txt [hunk 2]
```diff
@@ -237,10 +237,20 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
...
     )
 elseif (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
-    set_target_properties(Diligent-GraphicsEngineVk-shared PROPERTIES
-        # Disallow missing direct and indirect dependencies to enssure that .so is self-contained
-        LINK_FLAGS "-Wl,--no-undefined -Wl,--no-allow-shlib-undefined"
-    )
+    if(PLATFORM_WIN32)
+        file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/export.map
+            "{ global: *GetEngineFactoryVk*; local: *; };"
+        )
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineVulkan/CMakeLists.txt:243]: Same as above

**Comment by logzero** [comment]
My CMake skills are pretty basic. Appending to LINK_FLAGS seems to work as expected.

**Comment by TheMostDiligent** [comment]
I still don't quite understand what your changes do and why they were necessary, but they certainly fixed linking errors - thanks!

**Comment by TheMostDiligent** [comment]
As to xinput errors - they come from imgui, will think what to do about them

**Comment by TheMostDiligent** [comment]
I updated imgui to the latest version and it fixed the problem with XInput, but introduced another problem with `SetProcessDPIAware`:
```
imgui_impl_win32.cpp:408:5: error: 'SetProcessDPIAware' was not declared in this scope
```
I don't know why MinGW is complaining, this function seems to be [around since 2008](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setprocessdpiaware).

**Comment by TheMostDiligent** [comment]
So commenting out [this line](https://github.com/ocornut/imgui/blob/master/examples/imgui_impl_win32.cpp#L408) makes everything work. I opened an [issue vs DearImgui](https://github.com/ocornut/imgui/issues/3183).

**Comment by logzero** [comment]
> 
> 
> I still don't quite understand what your changes do and why they were necessary, but they certainly fixed linking errors - thanks!

My understanding is that the visibility attributes are a compiler thing, which the compiler has to communicate to the linker. I guess mingw does not support them correctly?

-Wl,--version-script on the other side is telling the linker explicitely what to do.

**Comment by logzero** [comment]
Btw I used APPEND because set_target_properties override previous settings. Wasn't sure if that was intended.

**Comment by TheMostDiligent** [comment]
> set_target_properties override previous settings

Thanks for pointing out at this, I honestly was 100% sure that it appends link flags.

**Pull request merged**